rpc/wrtc_rate_limiter.go

```diff
@@ -0,0 +1,179 @@
+) (*MongoDBRateLimiter, error) {
+	rateLimitColl := client.Database(mongodbWebRTCCallQueueDBName).Collection(MongoDBRateLimiterCollName)
+
+	ttlSeconds := int32((2 * config.Window).Seconds())
+	indexes := []mongo.IndexModel{
+		{
+			Keys: bson.D{{Key: "created_at", Value: 1}},
+			Options: &options.IndexOptions{
+				Name:               &mongodbWebRCRateLimiterTTLName,
+				ExpireAfterSeconds: &ttlSeconds,
+			},
+		},
+	}
+
+	if err := mongoutils.EnsureIndexes(ctx, rateLimitColl, indexes...); err != nil {
+		return nil, err
+	}
+
+	return &MongoDBRateLimiter{
+		rateLimitColl: rateLimitColl,
+		config:        config,
+		logger:        logger,
+	}, nil
+}
+
+// Allow checks if a request is within rate limits and records it atomically.
+// The filter only matches if the count of requests within the window are below the limit.
+// The update creates a new array that adds the current timestamp and removes old timestamps outside the window.
+// This prevents race conditions and keeps the requests array bounded.
+func (rl *MongoDBRateLimiter) Allow(ctx context.Context, key string) error {
+	// Ensure a document exists for the key to handle first request case since a $expr filter can't check for
+	// non-existence and create the document if it doesn't exist. We use $setOnInsert to only set fields if the
+	// document is being created and don't care about the result or error here since a duplicate key error is
+	// fine if another process created it first.
+	//nolint:errcheck
+	_, _ = rl.rateLimitColl.UpdateOne(ctx,
+		bson.M{"_id": key},
+		bson.M{"$setOnInsert": bson.M{
+			"requests":   bson.A{},
+			"created_at": "$$NOW",
+		}},
+		options.Update().SetUpsert(true))
+
+	// Filter: only match if recent request count < limit
+	filter := bson.M{
+		"_id": key,
+		"$expr": bson.M{
+			"$lt": bson.A{
+				bson.M{
+					"$size": bson.M{
+						"$filter": bson.M{
+							"input": "$requests",
+							"as":    "req",
+							"cond": bson.M{
+								"$gte": bson.A{
+									"$$req",
+									bson.M{
+										"$dateSubtract": bson.M{
+											"startDate": "$$NOW",
+											"unit":      "second",
+											"amount":    rl.config.Window.Seconds(),
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+				rl.config.MaxRequests,
+			},
+		},
+	}
+
+	update := bson.A{
+		bson.M{
+			"$set": bson.M{
+				"requests": bson.M{
+					"$concatArrays": bson.A{
+						bson.M{
+							"$filter": bson.M{
+								"input": "$requests",
+								"as":    "req",
+								"cond": bson.M{
+									"$gte": bson.A{
+										"$$req",
+										bson.M{
+											"$dateSubtract": bson.M{
+												"startDate": "$$NOW",
+												"unit":      "second",
+												"amount":    rl.config.Window.Seconds(),
+											},
+										},
+									},
+								},
+							},
+						},
+						bson.A{"$$NOW"},
+					},
+				},
+				"created_at": "$$NOW",
+			},
+		},
+	}
+
+	result := rl.rateLimitColl.FindOneAndUpdate(ctx, filter, update)
```

Comment:
This can just be an `Update` now, we're not doing anything with the found document.